### PR TITLE
Provide facility to test at runtime if an architecture is supported

### PR DESCRIPTION
**Pull request by serge-sans-paille**

### CODE REVIEW

**Comment by serge-sans-paille** [comment]
Green, at last. I should probably rename `visit` into `map` and provide a `fold` operator of sort.

**Comment by wolfv** [comment]
This is going to be really useful for mamba :)

**Comment by serge-sans-paille** [comment]
> This is going to be really useful for mamba :)

Nice! Why?

File: include/xsimd/config/xsimd_arch.hpp [hunk 1]
```diff
@@ -204,6 +379,11 @@ namespace xsimd
           static constexpr bool contains() {
             return detail::contains<Arch, Archs...>::value;
           }
+
+          template<class F>
+          static void visit(F&& f) {
```
Comment (JohanMabille) [inline include/xsimd/config/xsimd_arch.hpp:384]: I think `map`, or `for_each` would be a better name.

File: include/xsimd/config/xsimd_instruction_set.hpp [hunk 2]
```diff
@@ -21,7 +21,7 @@
 // minor number can be in [0, 99]
 // patch number can be in [0, 999999]
 #define XSIMD_VERSION_NUMBER(major, minor, patch) \
-    ((((major) % 100) * 10000000) + (((minor) % 100) * 100000) + ((patch) % 100000))
+    ((((major) % 100u) * 10000000u) + (((minor) % 100u) * 100000u) + ((patch) % 100000u))
```
Comment (JohanMabille) [inline include/xsimd/config/xsimd_instruction_set.hpp:24]: I don't understand why you use `unsigned int`here while you changed form `unsigned` to `int` in the different structures modeling the supported architectures.

**Review by JohanMabille — CHANGES_REQUESTED** [review]
Nitpicking: 4 spaces indent and opening curly braces go on a new line.

**Comment by serge-sans-paille** [comment]
@JohanMabille should be good now. I switched version to unsigned because it's easier to reason on unsigned, and there's no value in signed version anyway (?)

**Comment by JohanMabille** [comment]
Indeed, maybe I mixed things or read a single commit diff instead of the whole changes.

**Review by JohanMabille — APPROVED** [review]

**Comment by guyuqi** [comment]
It failed to build on:
`Arm64 gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0` :

```
/home/yuqi/xsimd/include/xsimd/memory/../config/xsimd_arch.hpp:502:77:   required from ‘decltype (((xsimd::arch::detail::dispatcher<F, ArchList>*)this)->xsimd::arch::detail::dispatcher<F, ArchList>::functor(xsimd::arch::default_{}, (forward<Tys>)(xsimd::arch::detail::dispatcher::operator()::args)...)) xsimd::arch::detail::dispatcher<F, ArchList>::operator()(Tys&& ...) [with Tys = {unsigned int (&)[17], int}; F = sum; ArchList = xsimd::arch::arch_list<xsimd::arch::neon64, xsimd::arch::neon>; decltype (((xsimd::arch::detail::dispatcher<F, ArchList>*)this)->xsimd::arch::detail::dispatcher<F, ArchList>::functor(xsimd::arch::default_{}, (forward<Tys>)(xsimd::arch::detail::dispatcher::operator()::args)...)) = unsigned int]’
/home/yuqi/xsimd/test/test_arch.cpp:72:39:   required from here
/home/yuqi/xsimd/test/test_arch.cpp:53:11: error: ‘batch acc’ has incomplete type
   53 |     batch acc(static_cast<T>(0));
      |           ^~~
/home/yuqi/xsimd/test/test_arch.cpp:54:29: error: incomplete type ‘batch’ {aka ‘xsimd::batch<unsigned int, 16>’} used in nested name specifier
   54 |     const unsigned n = size / batch::size * batch::size;
      |                        ~~~~~^~~~~~~~~~~~~
/home/yuqi/xsimd/test/test_arch.cpp:54:43: error: incomplete type ‘batch’ {aka ‘xsimd::batch<unsigned int, 16>’} used in nested name specifier
   54 |     const unsigned n = size / batch::size * batch::size;
      |                        ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/home/yuqi/xsimd/test/test_arch.cpp:55:35: error: incomplete type ‘batch’ {aka ‘xsimd::batch<unsigned int, 16>’} used in nested name specifier
   55 |     for(unsigned i = 0; i != n; i += batch::size)
      |                                 ~~^~~~~~~~~~~~~~
/home/yuqi/xsimd/test/test_arch.cpp:56:16: error: invalid use of incomplete type ‘using batch = xsimd::arch::batch<unsigned int, xsimd::arch::neon64>’ {aka ‘class xsimd::batch<unsigned int, 16>’}
   56 |         acc += batch(data + i);
      |                ^~~~~~~~~~~~~~~
In file included from /home/yuqi/xsimd/include/xsimd/memory/../config/xsimd_align.hpp:15,
                 from /home/yuqi/xsimd/include/xsimd/memory/xsimd_alignment.hpp:14,
                 from /home/yuqi/xsimd/include/xsimd/xsimd.hpp:14,
                 from /home/yuqi/xsimd/test/test_utils.hpp:20,
                 from /home/yuqi/xsimd/test/test_arch.cpp:14:
/home/yuqi/xsimd/include/xsimd/memory/../config/xsimd_arch.hpp:42:11: note: declaration of ‘using batch = xsimd::arch::batch<unsigned int, xsimd::arch::neon64>’ {aka ‘class xsimd::batch<unsigned int, 16>’}
   42 |     class batch;
      |           ^~~~~

```
@serge-sans-paille @JohanMabille

**Pull request merged**